src/Graphs/DirectedGraphBuilder.cs

```diff
@@ -0,0 +1,210 @@
+                if (predecessors.Count == 0)
+                {
+                    builder.sources.Add(element);
+                }
+            }
+
+            return builder;
+        }
+
+        public static IDirectedAcyclicGraph<T> EmptyGraph()
+        {
+            return new DirectedGraphBuilder<T>().CreateAcyclicGraphUnsafe();
+        }
+
+        private DirectedGraphBuilder() { }
+
+        public DirectedGraphBuilder<T> AddElement(T element)
+        {
+            if (element == null) throw new ArgumentNullException(nameof(element));
+
+            if (!elements.Add(element))
+            {
+                return this;
+            }
+            sources.Add(element);
+            directSuccessors.Add(element, new HashSet<T>());
+            directPredecessors.Add(element, new HashSet<T>());
+
+            return this;
+        }
+
+        public DirectedGraphBuilder<T> AddArrow(T from, T to)
+        {
+            if (from == null) throw new ArgumentNullException(nameof(from));
+            if (to == null) throw new ArgumentNullException(nameof(to));
+            if (!elements.Contains(from)) throw new ArgumentException("Element not in graph.", nameof(from));
+            if (!elements.Contains(to)) throw new ArgumentException("Element not in graph.", nameof(to));
+            if (from.Equals(to))
+            {
+                throw new InvalidOperationException("Cannot make self-edges in directed acyclic graph.");
+            }
+
+            directSuccessors[from].Add(to);
+            directPredecessors[to].Add(from);
+
+            if (sources.Contains(to))
+            {
+                sources.Remove(to);
+            }
+
+            return this;
+        }
+
+        public DirectedGraphBuilder<T> RemoveElement(T element)
+        {
+            if (element == null) throw new ArgumentNullException(nameof(element));
+
+            if (!elements.Remove(element))
+            {
+                return this;
+            }
+            sources.Remove(element);
+
+            foreach (var to in directSuccessors[element])
+            {
+                directPredecessors[to].Remove(element);
+                if (directPredecessors[to].Count == 0)
+                {
+                    sources.Add(to);
+                }
+            }
+            foreach (var from in directPredecessors[element])
+            {
+                directSuccessors[from].Remove(element);
+            }
+
+            return this;
+        }
+
+        public DirectedGraphBuilder<T> RemoveArrow(T from, T to)
+        {
+            if (from == null) throw new ArgumentNullException(nameof(from));
+            if (to == null) throw new ArgumentNullException(nameof(to));
+            if (!elements.Contains(from)) throw new ArgumentException("Element not in graph.", nameof(from));
+            if (!elements.Contains(to)) throw new ArgumentException("Element not in graph.", nameof(to));
```

Comment:
Do we need these two checks? Can't we just see if the edge is known?

I'm a bit confused why it will fail if the elements aren't known, but if the edge isn't known it's fine.